File: Problem72_Subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        if(nums == null || nums.length == 0)
            return result;
        result.add(new ArrayList<>());
        for(int n : nums) {
            int size = result.size();
            for(int i = 0; i < size; i++) {
                List<Integer> set = new ArrayList<Integer>(result.get(i));
                set.add(n);
                result.add(set);
            }
        }
        return result;
    }
}

File: Problem-72 Subsets.py
# 78. Subsets
# https://leetcode.com/problems/subsets/

# Time Complexiety: O(n*(2^n))
# Space Complexiety: O(n)

class Solution:
    def __init__(self):
        self.result = list()
        
    def _helper(self, nums, idx, path):
        self.result.append(path.copy())
        # Base
        if idx == len(nums):    
            return
        
        for i in range(idx, len(nums)):
            # Action
            path.append(nums[i])
            
            # Recursion
            self._helper(nums, i+1, path)
            
            # Backtrack
            path.pop()
        
    
    def subsets(self, nums):
        self._helper(nums, 0, [])
        return self.result
